ボウリングのスコア計算プログラムオブジェクト指向版のプログラムを作成

sjabcdefin opened the pull request:
ボウリングのスコア計算プログラムオブジェクト指向版のプログラムを作成いたしました。
ご確認お願いいたします。

sjabcdefin committed:
first commit

sjabcdefin committed:
Update: Change variable name and method name

変数名とメソッド名を変更

sjabcdefin committed:
Update: split point method

ポイントメソッドの行数が長かったため、ポイントメソッドを分割

maedana commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,29 @@
+
+  def spare?
+    @first_shot.score + @second_shot.score == 10
+  end
```
これだとストライクのときもtrueで判定されないですか? ストライクのときはfalseとなるようにしたいですね

sjabcdefin replied:
ストライクのときは、```false```になるように修正いたしました。
```
  def spare?
    !strike? && @first_shot.score + @second_shot.score == 10
  end
```

maedana commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,63 @@
+
+class Game
+  def initialize
+    @frames = split_into_frames
```
Frameクラスがあるわけですし、`@frames`には10個のFrameオブジェクトが入ることを期待しますね。工夫すればできそうですけどいかがですか?

sjabcdefin replied:
```@frames```に10個のFrameオブジェクトが入るように修正いたしました。

maedana commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,63 @@
+    point = 0
+    (0..9).each do |num|
+      frame, next_frame, after_next_frame = frame(num)
+      if frame.strike?
```
Frameのspare?メソッドで言及したことに関連しますけども、先にspareを判定したとしても機能するようにしたいですね。

sjabcdefin replied:
https://github.com/sjabcdefin/ruby-practices/pull/11#discussion_r1881620146

上記修正により、先にspareを判定したとしても機能するようになりました。

maedana commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,63 @@
+end
+
+game = Game.new
+puts game.point
```
クラスのファイルにはクラス定義だけを書きたいですね。gameクラスを利用する別のファイルから利用する形にしましょう。

sjabcdefin replied:
```bowling.rb```を作成し、このファイルからGameクラスを利用するようにいたしました。

maedana commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,63 @@
+      frame, next_frame, after_next_frame = frame(num)
+      if frame.strike?
+        point += frame.score + next_frame.first_shot.score + next_frame.second_shot.score
+        point += after_next_frame.first_shot.score if next_frame.first_shot.score == 10
```
ストライク判定は別の方法を用意してあるはずですよね。それを使いましょう

sjabcdefin replied:
```strike?```を使用するように修正いたしました。

maedana commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,29 @@
+      @first_shot.score,
+      @second_shot.score,
+      @third_shot.score
+    ].sum
```
この実装ですと、最終フレーム以外のときは基本点のみ、最終フレーム(3投ある)のときは基本点+ボーナスとなるわけですよね。一貫性がない感じが気になりますけど、この辺はどのように考えますか?

sjabcdefin replied:
```game.rb```でボーナス点を計算、```frame.rb```で基本点を計算として、最終フレーム以外のときも、基本点＋ボーナスで計算している認識でした。
ただ、スコア計算の責務が```Game```クラスと```Frame```クラスで分散していると思い、各フレームのスコア計算は、```Frame```クラスで実施するように修正いたしました。

ご指摘内容について、理解が不足していた場合申し訳ございません。🙇‍♀️
再度ご指摘いただければ幸いです。😌

maedana replied:
今回の対応方針でいいと思います。

> スコア計算の責務がGameクラスとFrameクラスで分散している

そうですね。そこまで求めなくてもいいかなと思ってましたが、よりよい対応になったと思います。

sjabcdefin committed:
Update: Modify the indicated items.

指摘事項を修正
- game.rbにはGameクラスのみを記載。bowling.rbを新規作成し、このファイルからgameクラスを利用するように修正。
- Frameクラスのspare?メソッドについて、ストライクのときはfalseを返すように修正。
- Gameクラスの@framesには10個のFrameオブジェクトが入るように修正。
- Frameクラスは、各フレームのスコア計算結果を返すように修正。

maedana commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,62 @@
+      base_score
+    end
+  end
+end
```
クラス外から利用しないメソッドはカプセル化の観点から`private`にしましょう。

sjabcdefin replied:
```base_score```, ```spare?```, ```final_frame?```, ```strike_bonus```, ```spare_bonus```メソッドを```private```にしました。
```strike?```メソッドについて、```public```にするか```protected```にするか迷いましたが、Frameクラス外では利用しないため、```protected```にしました。

maedana replied:
> strike?メソッド

private でいいんじゃないですかねー? protectedにした意図はなんでしょう?

sjabcdefin replied:
```strike_bonus?```メソッドで、```strike?```をレシーバを指定して呼び出しているため、```protected```としました。🤔🤔🤔
認識誤っていましたら、教えていただければ幸いです。🙇‍♀️

maedana replied:
失礼しました。ちゃんと読めてなかったですが、privateだとまずいですね。OKです。

maedana commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,36 @@
+      frame, next_frame, after_next_frame = @frames.slice(num, 3)
+      point += frame.total_score(next_frame, after_next_frame)
+    end
+    point
```
ここまでシンプルになったのであれば https://docs.ruby-lang.org/ja/latest/method/Array/i/sum.html 使って`point`というローカル変数はもはやなしで書いてもいいかなと思いました。ただ要対応とまでは思いませんので、ご対応は任意で大丈夫です。

sjabcdefin replied:
ローカル変数```point```を使用せず、```sum```を使用するように修正いたしました。

maedana commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,62 @@
+      base_score + spare_bonus(next_frame)
+    else
+      base_score
+    end
```
全ての分岐にbase_scoreがあるので、分岐の外に出せそうですね。すると少し分岐が減るかなと。いかがですか?

sjabcdefin replied:
```base_score```を分岐の外に配置しました。

maedana commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,62 @@
+      bonus = next_frame.first_shot.score + next_frame.second_shot.score
+      bonus += after_next_frame.first_shot.score if next_frame.strike? && after_next_frame
+      bonus
+    end
```
こういう場合早期returnにしてもいいかなと思います。好みもあるのでご対応は任意で大丈夫です。

```ruby
  return 0 if final_frame?

  bonus = ...
```
みたいなやつですね。

sjabcdefin replied:
早期```return```にするか迷っていました。修正いたしました。

maedana commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,62 @@
+  end
+
+  def spare?
+    !strike? && @first_shot.score + @second_shot.score == 10
```
`10`というのが2回出てくるので定数にまとめてもよいかなと思います。ただ、この程度であればご対応は任意で大丈夫かなと思います。

追記
`Shot`の10も含めると3箇所ですね

sjabcdefin replied:
bowling.rbに、```STRIKE_SCORE = 10```を定義し、```10```記載部分を```STRIKE_SCORE```に変更しました。

sjabcdefin committed:
Update:Using constant, Implement Encapsulation, Place base_score outside branch and Using early return.

指摘事項を修正
- bowling.rbに、STRIKE_SCORE = 10の定義を追加し、10をSTRIKE_SCOREに変更
- Gameクラスのpointメソッドについて、ローカル変数pointを削除し、sumメソッドを使用するように修正
- Frameクラスについて、カプセル化を実施
- Frameクラスtotal_scoreメソッド内のbase_scoreを分岐外に配置し、分岐をシンプルに修正
- Frameクラスのstrike_bonusメソッドについて、final_frameの場合、早期returnに変更
